cpu/nrf52/radio/nrf802154/nrf802154_radio.c

```diff
@@ -0,0 +1,703 @@
+ */
+static void _set_cca_thresh(uint8_t thresh)
+{
+    NRF_RADIO->CCACTRL &= ~RADIO_CCACTRL_CCAEDTHRES_Msk;
+    NRF_RADIO->CCACTRL |= thresh << RADIO_CCACTRL_CCAEDTHRES_Pos;
+}
+
+/**
+ * @brief   Convert from dBm to the internal representation, when the
+ *          radio operates as a IEEE802.15.4 transceiver.
+ */
+static inline uint8_t _dbm_to_ieee802154_hwval(int8_t dbm)
+{
+    return ((dbm - ED_RSSIOFFS) / ED_RSSISCALE);
+}
+
+static int set_cca_threshold(ieee802154_dev_t *dev, int8_t threshold)
+{
+    (void) dev;
+    _set_cca_thresh(_dbm_to_ieee802154_hwval(threshold));
+    return 0;
+}
+
+static void _set_txpower(int16_t txpower)
+{
+    if (txpower > 8) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Pos8dBm;
+    }
+    else if (txpower > 1) {
+        NRF_RADIO->TXPOWER = (uint32_t)txpower;
+    }
+    else if (txpower > -1) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_0dBm;
+    }
+    else if (txpower > -5) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg4dBm;
+    }
+    else if (txpower > -9) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg8dBm;
+    }
+    else if (txpower > -13) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg12dBm;
+    }
+    else if (txpower > -17) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg16dBm;
+    }
+    else if (txpower > -21) {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg20dBm;
+    }
+    else {
+        NRF_RADIO->TXPOWER = RADIO_TXPOWER_TXPOWER_Neg40dBm;
+    }
+}
+
+static void _disable(void)
+{
+    /* set device into DISABLED state */
+    if (NRF_RADIO->STATE != RADIO_STATE_STATE_Disabled) {
+        NRF_RADIO->EVENTS_DISABLED = 0;
+        NRF_RADIO->TASKS_DISABLE = 1;
+        DEBUG("[nrf802154] Device state: DISABLED\n");
+    }
+}
+
+static void _set_ifs_timer(bool lifs)
+{
+    uint8_t timeout;
+    ifs = true;
+    if (lifs) {
+        timeout = LIFS;
+    }
+    else {
+        timeout = SIFS;
+    }
+
+    timer_set(NRF802154_TIMER, MAC_TIMER_CHAN_IFS, timeout);
+    timer_start(NRF802154_TIMER);
+}
+
+static int _confirm_set_trx_state(ieee802154_dev_t *dev)
+{
+    (void) dev;
+    int radio_state = NRF_RADIO->STATE;
+    if (radio_state == RADIO_STATE_STATE_TxRu || radio_state == RADIO_STATE_STATE_RxRu ||
+        radio_state == RADIO_STATE_STATE_TxDisable || radio_state == RADIO_STATE_STATE_RxDisable) {
+        return -EAGAIN;
+    }
+    return 0;
+}
+
+static int _request_set_trx_state(ieee802154_dev_t *dev, ieee802154_trx_state_t state)
+{
+    (void) dev;
+
+    if (_state != STATE_IDLE && _state != STATE_RX) {
+        return -EBUSY;
+    }
+
+    _disable();
+
+    /* This will take in worst case 21 us */
+    while (NRF_RADIO->STATE != RADIO_STATE_STATE_Disabled);
+
+    switch(state) {
+        case IEEE802154_TRX_STATE_TRX_OFF:
+            _state = STATE_IDLE;
+            break;
+        case IEEE802154_TRX_STATE_RX_ON:
+            NRF_RADIO->PACKETPTR = (uint32_t)rxbuf;
```

Comment:
That's a neat feature.
Would have been cool if we could just use the buffer provided by the upper layer here.

Reply:
It's really nice. We could do that if the buffer is static.